census/census.go

```diff
@@ -99,20 +141,63 @@ func (m *Manager) Init(storageDir, rootAuthPubKey string) error {
 			cns.RootKey = rootAuthPubKey
 		}
 		m.Census = cns
-		if err := os.WriteFile(filepath.Clean(nsConfig), []byte(""), 0o600); err != nil {
+		if err := os.WriteFile(nsConfig, []byte(""), 0o600); err != nil {
 			return err
 		}
 		return m.save()
 	}
 
-	jsonBytes, err := os.ReadFile(filepath.Clean(nsConfig))
+	jsonBytes, err := os.ReadFile(nsConfig)
 	if err != nil {
 		return err
 	}
-	if err := json.Unmarshal(jsonBytes, &m.Census); err != nil {
+	var namespaces Namespaces
+	if err := json.Unmarshal(jsonBytes, &namespaces); err != nil {
 		log.Warn("could not unmarshal json config file, probably empty. Skipping")
 		return nil
 	}
+
+	// Copy censuses of the expected version into m.Census and remove old ones
+	m.Census.RootKey = namespaces.RootKey
+	cleanOldCensues := false
+	now := time.Now()
+	for _, v := range namespaces.Namespaces {
+		switch v.Version {
+		case 0:
+			// Version 0 are graviton censuses.  These censuses are
+			// no longer supported and belong to legacy processes.
+			cleanOldCensues = true
+			outdatedDir := filepath.Join(storageDir,
+				fmt.Sprintf("outdated.%v", now.Format(time.RFC3339)), "v0")
+			if err := os.MkdirAll(outdatedDir, 0o700); err != nil {
+				return err
+			}
+			log.Infof("Moving outdated census %v to %v", v.Name, outdatedDir)
+			dir := strings.Split(v.Name, string(os.PathSeparator))[0]
+			src := filepath.Join(storageDir, dir)
+			dst := filepath.Join(outdatedDir, dir)
+			if err := os.Rename(src, dst); err != nil {
```

Comment:
do you expect both to be in the same filesystem? because if not, this rename might fail

Reply:
I don't expect the dst and src to be in a different filesystem, because both dst and src share the same ancestor directory (storageDir).  It could happen, but it would be a strange situation.  Nevertheless, I haven't found any function in the standard library that does the same as `mv`; and I don't think it's worth adding a dependency to have this function as it would be a very strange case.

Reply:
Yep agreed. I mentioned this in case the user could configure src and dst to end up in separate filesystems in a reasonable way, but it seems to not be the case.